Blog post for CP ecosystem (#53)
jump-dev/jump-dev.github.io: _posts/2021-09-18-cp-0-6.md

---
layout: post
title: "ConstraintProgrammingExtensions 0.6 is released"
date: 2021-09-18
categories: [releases]
permalink: /blog/cpe-0.6-release/
---

> JuMP and MathOptInterface are mostly oriented towards mathematical optimisation. 

Reviewer: ```suggestion
JuMP and MathOptInterface are oriented towards mathematical optimization. 
```

However, their interface is amenable to other kinds of formalism, including 
constraint programming (CP). Bridging the gap between MOI and constraint programming
is the goal of the 
[ConstraintProgrammingExtensions.jl](https://github.com/dourouc05/ConstraintProgrammingExtensions.jl)
package. 

Discussions for support of constraint programming in the JuMP ecosystem started
[in 2019](https://github.com/jump-dev/JuMP.jl/issues/2014). The original scope
was quite limited, as it only planned to propose some modeling capabilities to 
JuMP. However, quite soon, it morphed into a larger proposal of a full 

> [constraint programming environment](https://github.com/jump-dev/JuMP.jl/issues/2227)

Reviewer: ```suggestion
[constraint programming environment](https://github.com/jump-dev/JuMP.jl/issues/2227),
```

accessible from the same API as JuMP. It also meant giving access to CP solvers. 

> Mathematical optimisation models are only based on equality and inequality constraints, 

Reviewer: This isn't quite correct in MOI. How about 

```
Traditional mathematical optimization models focus on (in)equality constraints,
```

typically linear or convex, with an objective function. Constraint programming 
allows more variety in the kind of constraints that are supported, with a 
focus on discrete structure: the best-known constraint is probably `all_different`, 
which models the fact that several integer-valued variables should take different

> values (no two variables are allowed to have the same value). Extremely efficient

Reviewer: ```suggestion
values (no two variables are allowed to have the same value). Efficient
```

  It supports the FlatZinc file format, a standard in constraint programming to exchange
  models (similar to the LP and MPS formats for linear programming), both to read and 
  write models, but also to communicate with solvers. It has recently been updated
  to be compatible with MOI 0.10.
  
- Several solvers are accessible using this API: the pure-Julia 
  [ConstraintSolver.jl](https://github.com/Wikunia/ConstraintSolver.jl/), the industry standard 
  [IBM CPLEX CP Optimizer](https://github.com/dourouc05/CPLEXCP.jl/), and the open-source
  [Chuffed](https://github.com/dourouc05/Chuffed.jl/).
  
- Other pure-Julia CP solvers have been released, but they do not yet use the same API:
  [JuliaConstraints](https://github.com/JuliaConstraints) and 
  [SeaPerl](https://github.com/corail-research/SeaPearl.jl). 
  
This is not the end of the journey. Although MOI and JuMP are nearing their 1.0 
release, the CP ecosystem has still a long way to go. In the near term, more
modelling features should be added as extensions to JuMP and more solvers should
be made accessible. In the longer term, CP modelling will also benefit from 

> the new generation of nonlinear support. 

Reviewer: You should add you name to this post. Either at the end, or at the top? Something like:
```
# ConstraintProgrammingExtensions 0.6 is released

_Author: Thibaut Cuvelier (@dourouc05)_
```

---
layout: post
title: "An update on constraint programming in JuMP"
date: 2021-09-18
categories: [releases]
permalink: /blog/constraint-programming-update/
---

_Author: Thibaut Cuvelier (@dourouc05)_

JuMP and MathOptInterface are oriented towards traditional mathematical
optimization, encompassing problem classes such mixed-integer linear programs
and conic optimization. However, the MathOptInterface API is amenable to other

> kinds of formalism, including constraint programming.

Reviewer: ```suggestion
kinds of formalism, including [constraint programming](https://en.wikipedia.org/wiki/Constraint_programming).
```

MathOptInterface.

Discussions about constraint programming support in the JuMP ecosystem started
[in 2019](https://github.com/jump-dev/JuMP.jl/issues/2014). The original scope
was quite limited, as it only planned to propose some modeling capabilities to
JuMP. However, quite soon, it morphed into a larger proposal of a full
[constraint programming environment](https://github.com/jump-dev/JuMP.jl/issues/2227),
accessible from the same API as JuMP, including wrapping various specialized
constraint programming solvers.

In the last few months, the constraint programming ecosystem in JuMP has made
tremendous progress. Highlights include:

- The v0.6 release of [ConstraintProgrammingExtensions.jl](https://github.com/dourouc05/ConstraintProgrammingExtensions.jl).
  Highlights of the ConstraintProgrammingExtensions.jl package include:
  * set definitions for [dozens of constraint types](http://tcuvelier.be/ConstraintProgrammingExtensions.jl/dev/reference/sets/)
  * a number of [bridges](http://tcuvelier.be/ConstraintProgrammingExtensions.jl/dev/reference/bridges_sets/),
    including many bridges to convert constraint programming models into
    standard mixed-integer programming models

>   * Support for the FlatZinc file format, a standard in constraint programming

Reviewer: ```suggestion
  * Support for the [FlatZinc file format](https://www.minizinc.org/doc-2.4.3/en/fzn-spec.html), a standard in constraint programming
```

    to exchange models (similar to the LP and MPS formats for linear programming)

- Interfacing several solvers through ConstraintProgrammingExtensions.jl. These

>   include

Reviewer: ```suggestion
  include:
```

  * The pure-Julia [ConstraintSolver.jl](https://github.com/Wikunia/ConstraintSolver.jl/)
  * The industry standard [IBM CPLEX CP Optimizer](https://github.com/dourouc05/CPLEXCP.jl/)
  * The open-source [Chuffed](https://github.com/dourouc05/Chuffed.jl/)
  Other pure-Julia constraint solvers have been released, but they do not yet
  use ConstraintProgrammingExtensions API:
  * [JuliaConstraints](https://github.com/JuliaConstraints)

>   * [SeaPerl](https://github.com/corail-research/SeaPearl.jl)

Reviewer: ```suggestion
  * [SeaPearl](https://github.com/corail-research/SeaPearl.jl)
```

- A session on constraint programming talks at JuMP-dev 2021
  * [The design of MiniZinc](https://www.youtube.com/watch?v=tHi0Jd7vgVA)
  * [ConstraintProgrammingExtensions.jl](https://www.youtube.com/watch?v=B0lO6HdlFAw&list=PLP8iPy9hna6Q343_8sSq4f306VGLW4TLK&index=104)
  * [ConstraintSolver.jl](https://www.youtube.com/watch?v=FXqQc8bi_vw&list=PLP8iPy9hna6Q343_8sSq4f306VGLW4TLK&index=101)
  * [JuliaConstraints](https://www.youtube.com/watch?v=G4siuvNMj0c&list=PLP8iPy9hna6Q343_8sSq4f306VGLW4TLK&index=4)

> This is not the end of the journey. Although MathOptInterface and JuMP are nearing

Reviewer: ```suggestion
This is the start of our journey towards constraint programming in JuMP. Although MathOptInterface and JuMP are nearing
```

---
layout: post
title: "An update on constraint programming in JuMP"
date: 2021-09-18
categories: [releases]
permalink: /blog/constraint-programming-update/
---

_Author: Thibaut Cuvelier (@dourouc05)_

JuMP and MathOptInterface are oriented towards traditional mathematical

> optimization, encompassing problem classes such mixed-integer linear programs

Reviewer 2: "such" -> "such as"

  * The open-source [Chuffed](https://github.com/dourouc05/Chuffed.jl/)
  
  Other pure-Julia constraint solvers have been released, but they do not yet
  use ConstraintProgrammingExtensions API:
  * [JuliaConstraints](https://github.com/JuliaConstraints)
  * [SeaPearl](https://github.com/corail-research/SeaPearl.jl)

- A session on constraint programming talks at JuMP-dev 2021
  * [The design of MiniZinc](https://www.youtube.com/watch?v=tHi0Jd7vgVA)
  * [ConstraintProgrammingExtensions.jl](https://www.youtube.com/watch?v=B0lO6HdlFAw&list=PLP8iPy9hna6Q343_8sSq4f306VGLW4TLK&index=104)
  * [ConstraintSolver.jl](https://www.youtube.com/watch?v=FXqQc8bi_vw&list=PLP8iPy9hna6Q343_8sSq4f306VGLW4TLK&index=101)
  * [JuliaConstraints](https://www.youtube.com/watch?v=G4siuvNMj0c&list=PLP8iPy9hna6Q343_8sSq4f306VGLW4TLK&index=4)

This is the start of our journey towards constraint programming in JuMP. Although MathOptInterface and JuMP are nearing
their 1.0 release, the constraint programming ecosystem has still a long way to go.
In the near term, more modelling features should be added as extensions to JuMP and
more solvers should be made accessible. In the longer term, constraint

> programming modelling will also benefit from the next generation of nonlinear

Reviewer 2: Maybe reference https://github.com/jump-dev/MathOptInterface.jl/issues/846?
